Tests/AuthenticationInterceptorTests.swift

```diff
@@ -0,0 +1,663 @@
+            self.requiresRefresh = requiresRefresh
+        }
+    }
+
+    enum OAuthError: Error {
+        case refreshNetworkFailure
+    }
+
+    class OAuthAuthenticator: Authenticator {
+        private(set) var applyCount = 0
+        private(set) var refreshCount = 0
+        private(set) var didRequestFailDueToAuthErrorCount = 0
+        private(set) var isRequestAuthenticatedWithCredentialCount = 0
+
+        let refreshResult: Result<OAuthCredential, Error>?
+        let lock = NSLock()
+
+        init(refreshResult: Result<OAuthCredential, Error>? = nil) {
+            self.refreshResult = refreshResult
+        }
+
+        func apply(_ credential: OAuthCredential, to urlRequest: inout URLRequest) {
+            lock.lock() ; defer { lock.unlock() }
+
+            applyCount += 1
+
+            let authorization = HTTPHeader.authorization(bearerToken: credential.accessToken)
+            urlRequest.addValue(authorization.value, forHTTPHeaderField: authorization.name)
+        }
+
+        func refresh(_ credential: OAuthCredential, completion: @escaping (Result<OAuthCredential, Error>) -> Void) {
+            lock.lock() ; defer { lock.unlock() }
+
+            refreshCount += 1
+
+            let refreshResult = self.refreshResult ?? .success(
+                OAuthCredential(
+                    accessToken: "a\(refreshCount)",
+                    refreshToken: "a\(refreshCount)",
+                    userID: "u1",
+                    expiration: Date()
+                )
+            )
+
+            DispatchQueue.global(qos: .utility).asyncAfter(deadline: .now() + 0.1) { completion(refreshResult) }
+        }
+
+        func didRequest(_ urlRequest: URLRequest, with response: HTTPURLResponse, failDueToAuthenticationError error: Error) -> Bool {
+            lock.lock() ; defer { lock.unlock() }
+
+            didRequestFailDueToAuthErrorCount += 1
+
+            return response.statusCode == 401
+        }
+
+        func isRequest(_ urlRequest: URLRequest, authenticatedWith credential: OAuthCredential) -> Bool {
+            lock.lock() ; defer { lock.unlock() }
+
+            isRequestAuthenticatedWithCredentialCount += 1
+
+            let bearerToken = HTTPHeader.authorization(bearerToken: credential.accessToken).value
+
+            return urlRequest.value(forHTTPHeaderField: "Authorization") == bearerToken
+        }
+    }
+
+    class PathAdapter: RequestAdapter {
+        var paths: [String]
+
+        init(paths: [String]) {
+            self.paths = paths
+        }
+
+        func adapt(_ urlRequest: URLRequest, for session: Session, completion: @escaping (Result<URLRequest, Error>) -> Void) {
+            var request = urlRequest
+
+            var urlComponents = URLComponents(string: request.url!.absoluteString)!
+            urlComponents.path = paths.removeFirst()
+
+            request.url = urlComponents.url
+
+            DispatchQueue.main.async { completion(.success(request)) }
+        }
+    }
```

Comment:
This is a fun little class that allows me to modify the path of the request as it is retried.